File: lua/exercises.lua
local exercises = {}

-- this function is to calculate the absolute value
local function abs(x)
  return x < 0 and -x or x
end

-- Implement change function
function exercises.change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  elseif amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination], remaining = remaining // denomination, remaining % denomination
  end
  return counts
end

-- Implement firstThenLowerCase function
function exercises.firstThenLowerCase(strings, predicate)
  for _, s in ipairs(strings) do
    if predicate(s) then
      return s:lower()
    end
  end
  return nil
end

-- Implement powersGenerator coroutine
function exercises.powersGenerator(base, limit)
  return coroutine.create(function()
    local exponent = 0
    while true do
      local power = base ^ exponent
      if power > limit then break end
      coroutine.yield(power)
      exponent = exponent + 1
    end
  end)
end

-- Implement say function
function exercises.say(word)
  local words = {}
  local function innerSay(newWord)
    if not newWord then
      return table.concat(words, " ")
    end
    table.insert(words, newWord)
    return innerSay
  end
  return innerSay(word)
end

-- Implement meaningfulLineCount function
function exercises.meaningful_line_count(filename)
  local file = io.open(filename, "r")
  if not file then error("No such file") end
  local count = 0
  for line in file:lines() do
    if line:match("%S") then count = count + 1 end
  end
  file:close()
  return count
end

-- Implement Quaternion class
exercises.Quaternion = {}
exercises.Quaternion.__index = exercises.Quaternion

function exercises.Quaternion.new(a, b, c, d)
  return setmetatable({a = a, b = b, c = c, d = d}, exercises.Quaternion)
end

function exercises.Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end

function exercises.Quaternion:__add(q)
  return exercises.Quaternion.new(
    self.a + q.a, self.b + q.b, self.c + q.c, self.d + q.d
  )
end

function exercises.Quaternion:__mul(q)
  return exercises.Quaternion.new(
    self.a * q.a - self.b * q.b - self.c * q.c - self.d * q.d,
    self.a * q.b + self.b * q.a + self.c * q.d - self.d * q.c,
    self.a * q.c - self.b * q.d + self.c * q.a + self.d * q.b,
    self.a * q.d + self.b * q.c - self.c * q.b + self.d * q.a
  )
end

function exercises.Quaternion:conjugate()
  return exercises.Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

function exercises.Quaternion:__tostring()
  return string.format("%g+%gi+%gj+%gk", self.a, self.b, self.c, self.d)
end

function exercises.Quaternion:__eq(q)
  return abs(self.a - q.a) < 1e-9 and
         abs(self.b - q.b) < 1e-9 and
         abs(self.c - q.c) < 1e-9 and
         abs(self.d - q.d) < 1e-9
end

return exercises

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable

# this function to calculate change 
def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts

# this function is to find the first string matching the predicate, then return it in lowercase
def first_then_lower_case(strings, predicate, /):
    for string in strings:
        if predicate(string):
            return string.lower()
    return None

# this generator function is to yield powers of a base up to a given limit
def powers_generator(base: int, limit: int):
    power = 0
    while base ** power <= limit:
        yield base ** power
        power += 1

# this function is to create a chainable sentence builder
def say(word=None):
    if not hasattr(say, "words"):
        say.words = []

    def inner(word=None):
        if word is None:
            result = ' '.join(say.words)
            say.words = []  # Reset list after getting the result
            return result
        else:
            say.words.append(word)  # directly add the word
            return inner

    return inner(word)

# this function is to count meaningful lines in a file 
def meaningful_line_count(filename: str) -> int:
    try:
        with open(filename, 'r') as file:
            count = 0
            for line in file:
                line = line.strip()
                if line and not line.startswith('#'):
                    count += 1
        return count
    except FileNotFoundError:
        raise

@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float

    def __add__(self, other):
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)

    def __mul__(self, other):
        return Quaternion(
            self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d,
            self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c,
            self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b,
            self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
        )

    @property
    def coefficients(self):
        return [self.a, self.b, self.c, self.d]

    @property
    def conjugate(self):
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __eq__(self, other):
        return isinstance(other, Quaternion) and self.coefficients == other.coefficients

    def __str__(self):
        parts = []

        # this to format the real section
        if self.a != 0:
            real_part = f"{self.a:.6g}"
            parts.append(real_part)

        # this to format the unreal section
        for value, symbol in [(self.b, 'i'), (self.c, 'j'), (self.d, 'k')]:
            if value != 0:
                abs_value = abs(value)
                part = symbol if abs_value == 1 else f"{abs_value:.6g}{symbol}"
                sign = '-' if value < 0 else '+' if parts else ''
                parts.append(f"{sign}{part}")

        # to combine parts and handle special cases
        result = "".join(parts)
        if result == "1":
            result = "1.0"
        elif result == "-1":
            result = "-1.0"
        if result.startswith('+'):
            result = result[1:]
        result = result.replace('+-', '-').replace(' ', '')

        return result if result else "0"
